spec.bs

```diff
@@ -1226,6 +1254,282 @@ On the other hand, methods for getting data from the [=shared storage database=]
+
+### [=Shared-Storage-Write=] Response Header ### {#response-header}
+
+  This specification defines a <dfn>Shared-Storage-Write</dfn> HTTP [=/response=] [=header=].
+
+  The [=Shared-Storage-Write=] [=/response=] [=header=] is a [=list structured header=], for which the [=parameters|parameterized=] [=structured header/items=] can have one of the following values, expressed as [=item/token=], [=item/string=], or [=item/byte sequence=]:
+
+    * `set`
+        - required [=parameters=]:
+            * key ([=item/Token=], [=item/String=], or [=item/Byte Sequence=])
+            * value ([=item/Token=], [=item/String=], or [=item/Byte Sequence=])
+        - optional [=parameters|parameter=]:
+            * ignore_if_present ([=item/Boolean=])
+    * `append`
+        - required [=parameters=]:
+            * key ([=item/Token=], [=item/String=], or [=item/Byte Sequence=])
+            * value ([=item/Token=], [=item/String=], or [=item/Byte Sequence=])
+    * `delete`
+        - required [=parameters|parameter=]:
+            * key ([=item/Token=], [=item/String=], or [=item/Byte Sequence=])
+    * `clear`
+
+<span class=todo>Add examples.</span>
+
+## Shared Storage Fetch-Related Algorithms ## {#ss-fetch-algo}
+
+  <div algorithm>
+    To <dfn>determine whether a request can use shared storage</dfn>, given a [=/request=] |request|, perform the following steps:
+
+    1. Set |window| to |request|’s [=request/window=].
+    1. If |window| is not an [=environment settings object=] whose [=global object=] is a {{Window}}, return false.
+    1. If the result of running [=determine whether shared storage is allowed=] for |window| is false, return false.
+    1. Let |document| be |window|'s [=global object=]'s [=associated document=].
+    1. Return the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage=]", |document|, and |request|'s [=request/current URL=]'s [=url/origin=].
+
+    Issue: The [=determine whether a request can use shared storage=] algorithm needs to take into account "opt-in features", as articulated in <a href="https://github.com/w3c/webappsec-permissions-policy/pull/499">https://github.com/w3c/webappsec-permissions-policy/pull/499</a>.
+  </div>
+
+
+  <div algorithm>
+    To <dfn>append or modify a Shared-Storage-Writable request header</dfn>, given a [=/request=] |request|, perform the following steps:
+
+    1. If |request|'s [=request/shared storage writable=] is not true, then abort these steps and return.
+    1. If the result of running [=determine whether a request can use shared storage=] on |request| is false, perform the following steps:
+        1. [=header list/Delete=] "[=Shared-Storage-Writable=]" from |request|'s [=request/header list=].
+        1. Set |request|'s [=request/shared storage writable=] to false.
```

Comment:
I think we wouldn't want to set `shared storage writable` to false, because it would deny all future redirects. For example, for a redirect chain of A->B->C, C could still be eligible even if B isn't eligible.

Reply:
Interesting. The current implementation would disallow such a chain. Once a request gets marked ineligible due to an ineligible origin in the redirect chain, it cannot currently become eligible again.

See [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/loader/frame_fetch_context.cc;l=360) for `fetch()` and images, and [here](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_request.cc;l=5111) for iframes.

We could change this in both the implementation and spec, but of course they should match. What do you think?

Reply:
Ah I didn't catch this in code review. Yes I think we should change this in both the implementation and spec, so that each redirect is checked independently.